### PR TITLE
bin/ruby-build: support DESTDIR setting

### LINKED ISSUE #1415: DESTDIR setting fails
Hi,
I am trying to build Ruby for packaging with fpm, with these two steps:

PREFIX=/usr/local/ruby ./install.sh
MAKE_INSTALL_OPTS="DESTDIR=/tmp/ruby" ruby-build "2.7.0" "/usr/local/ruby"

This installs Ruby to e.g. `/tmp/usr/local/ruby/bin/ruby`, so far so good. However, `RUBY_BIN` is set to `/usr/local/ruby/bin`, not `/tmp/usr/local/ruby/bin`, so anything that calls it, for example `verify_openssl`, also fails, because the binary is in the wrong place.

It would be nice to have a global DESTDIR setting that modifies RUBY_BIN (which appears to be unsettable from the outside) and also modifies MAKE_INSTALL_OPTS.

### PR DESCRIPTION
**Pull request by kevinburke1**
We can't just use MAKE_INSTALL_OPTS because DESTDIR modifies the
location of the installed ruby binary, so anything that used the old
RUBY_BIN also failed - it needed to be modified as well.

Instead introduce a new variable that edits both RUBY_BIN and
MAKE_INSTALL_OPTS if set.

Also edit LDFLAGS so we can find libruby.so in the DESTDIR for
post-install checks.

Fixes #1415.
Updates #413.

### CODE REVIEW

**Comment by kevinburkemeter** [comment]
I've verified that this works on Linux installing to a temporary directory.

File: bin/ruby-build [hunk 1]
```diff
@@ -1374,11 +1379,21 @@ if [ -n "$VERBOSE" ]; then
...
+    export CPPFLAGS="-I${DESTDIR}${PREFIX_PATH}/include ${CPPFLAGS}"
+else
+    export LDFLAGS="-L${PREFIX_PATH}/lib ${LDFLAGS}"
+    export CPPFLAGS="-I${PREFIX_PATH}/include ${CPPFLAGS}"
+fi
 
 unset RUBYOPT
-unset RUBYLIB
+if [ -n "${DESTDIR}" ]; then
+    export RUBYLIB="${DESTDIR}${PREFIX_PATH}/lib/ruby/${ARGUMENTS[0]}:${DESTDIR}${PREFIX_PATH}/lib/ruby/${ARGUMENTS[0]}/$(uname -m)-$(uname | tr '[:upper:]' '[:lower:]')"
```
[the remark this answers is not in the document]
Comment (kevinburkemeter) [reply bin/ruby-build]: I'm also not super happy about the use of ARGUMENTS[0], I can pull this into a helper variable.

[the remark this answers is not in the document]
Comment (mislav) [reply bin/ruby-build]: Hmm, these additions add a moderate complexity to the code, and I don't believe they even cover the full range of where `PREFIX_PATH` is being used right now to reference locations written on disk after the compile phase 🤔 

I wonder whether this is feasible for us to support. Note that we should also try to reasonably support DESTDIR in rubies other than MRI, such as JRuby, Rubinius, and others.

Or, perhaps we should put our foot down and suggest that anybody who needs to make an advanced build of a MRI version download the tarball and pass all necessary make options themselves, effectively skipping ruby-build.

How are you using DESTDIR, by the way?

**Comment by kevinburkemeter** [comment]
Here's the script I have currently...

```bash
install_ruby() {
    set -x
    local -r numcpu=$(grep -c ^processor /proc/cpuinfo)
    if [[ "${numcpu}" == '' ]]; then
        numcpu='1'
    fi
    pushd /tmp
        git clone git://github.com/meterup/ruby-build.git
        pushd ruby-build
            PREFIX="${2}" "./install.sh"
            # https://github.com/rbenv/ruby-build#custom-build-configuration
            MAKE_OPTS="-j${numcpu}" DESTDIR="${3}" ruby-build "${1}" "${2}"
        popd
    popd
}
```

**Comment by kevinburkemeter** [comment]
Another option would be to tell people using `DESTDIR` that they're on their own and skip the post-install `RUBY_BIN` verification steps.

**Comment by mislav** [comment]
> Another option would be to tell people using `DESTDIR` that they're on their own and skip the post-install `RUBY_BIN` verification steps.

That sounds like a pretty good idea! Could you tell how much work would be involved in that compared to this PR?

**Comment by kevinburkemeter** [comment]
Oh... basically it would be to just add a bit before every invocation of RUBY_BIN (for example verify_config) that skips running the command if DESTDIR is set.

**Comment by kevinburkemeter** [comment]
Another issue with this patch I am noticing now that 2.7.1 has been released:

This patch sets RUBYLIB to e.g.

```
$DESTDIR/usr/local/lib/ruby/2.7.1:$DESTDIR/usr/local/lib/ruby/2.7.1/x86_64-linux
```

However, Ruby installs the directories for 2.7.1 to `$DESTDIR/usr/local/lib/ruby/2.7.0`. So this patch is looking in the wrong place.

This also unfortunately means we need some kind of logic for getting the base patch version (2.7.0) from the updated patch version (2.7.1), which I don't think we currently have in the bash script.

**Comment by kevinburke** [comment]
The 2.7.1 issue should be fixed now.

**Comment by mislav** [comment]
Thank you for your contribution, but see https://github.com/rbenv/ruby-build/issues/1415#issuecomment-1257277411 🙇

**Pull request closed without merging**